File: apps/go/internal/chapter01/kvstdata_test.go
package chapter01_test

import (
	"bytes"
	"math"
	"os"
	"path/filepath"
	"testing"

	"github.com/k2works/getting-started-machine-learning/apps/go/internal/chapter01"
	"github.com/k2works/getting-started-machine-learning/apps/go/internal/dataset"
)

// requireData は学習データが無ければテストを飛ばす。
func requireData(t *testing.T) string {
	t.Helper()

	csvFile := filepath.Join(dataset.Current(), "KvsT.csv")
	if _, err := os.Stat(csvFile); err != nil {
		t.Skip("学習データ KvsT.csv が配置されていない（gulp data:setup）")
	}

	return csvFile
}

func TestKvsTData(t *testing.T) {
	t.Parallel()

	t.Run("実データから 19 人分を読み込む", func(t *testing.T) {
		t.Parallel()

		people, err := chapter01.LoadPeople(requireData(t))
		if err != nil {
			t.Fatalf("LoadPeople() でエラー: %v", err)
		}

		if want := 19; len(people) != want {
			t.Errorf("件数 = %d, want %d", len(people), want)
		}
	})

	t.Run("ルールによる判定の正解率を実データで計算する", func(t *testing.T) {
		t.Parallel()

		people, err := chapter01.LoadPeople(requireData(t))
		if err != nil {
			t.Fatalf("LoadPeople() でエラー: %v", err)
		}

		features, labels := chapter01.SplitFeaturesAndLabels(people)

		predictions := make([]string, len(features))
		for i, feature := range features {
			predictions[i] = chapter01.PredictByRule(feature)
		}

		got, err := chapter01.Accuracy(predictions, labels)
		if err != nil {
			t.Fatalf("Accuracy() でエラー: %v", err)
		}

		if want := 14.0 / 19; math.Abs(got-want) > 1e-12 {
			t.Errorf("正解率 = %v, want %v", got, want)
		}
	})

	t.Run("実行するとデータ件数と正解率を表示する", func(t *testing.T) {
		t.Parallel()

		_ = requireData(t)

		var out bytes.Buffer
		if err := chapter01.Run(&out); err != nil {
			t.Fatalf("Run() でエラー: %v", err)
		}

		want := "データ件数: 19\nルールによる判定の正解率: 0.7368\n"
		if out.String() != want {
			t.Errorf("出力 = %q, want %q", out.String(), want)
		}
	})
}

File: apps/ruby/test/kvst_data_test.rb
# frozen_string_literal: true

require "test_helper"

# 実データ（KvsT.csv）を使うテスト。学習データが無ければスキップする。
class KvsTDataTest < Minitest::Test
  # module_function のモジュールを include すると、章の run が Minitest の run を上書きするので、
  # include せずに定数と関数をモジュールから呼ぶ。
  C = GettingStartedMl::Chapter01

  def csv_file
    path = File.join(GettingStartedMl::Dataset.dir, "KvsT.csv")
    skip "学習データ KvsT.csv が配置されていない（gulp data:setup）のでスキップする" unless File.exist?(path)

    path
  end

  def test_実データを読み込んで正解率を求める
    people = C.load_people(csv_file)

    assert_equal 19, people.size

    x, t = C.split_features_and_labels(people)
    predictions = x.map { |features| C.predict_by_rule(features) }

    assert_in_delta 14.0 / 19, C.accuracy(predictions, t), 1e-12
  end
end
